dyn_load.c

```diff
@@ -480,18 +480,56 @@ STATIC int GC_register_dynlib_callback(struct dl_phdr_info * info,
 
   p = info->dlpi_phdr;
   for( i = 0; i < (int)info->dlpi_phnum; i++, p++ ) {
-    switch( p->p_type ) {
+    if( p->p_type == PT_LOAD) {
+      GC_has_static_roots_func callback = GC_has_static_roots;
+      if( !(p->p_flags & PF_W) ) continue;
+      start = ((char *)(p->p_vaddr)) + info->dlpi_addr;
+      end = start + p->p_memsz;
+
+      if (callback != 0 && callback(info->dlpi_name, start, p->p_memsz))
```

Comment:
Sorry, I should have tested this before commiting. I've tested now, and when using "callback" instead of "!callback", my program crashes almost immediately.